core/image/jpeg/jpeg.odin

```diff
@@ -0,0 +1,927 @@
+					}
+
+					quantization_table_idx := compress.read_u8(ctx) or_return
+
+					if quantization_table_idx < 0 || quantization_table_idx > 3 {
+						return img, .Invalid_Quantization_Table_Index
+					}
+
+					color_components[id].quantization_table_idx = quantization_table_idx
+					color_components[id].v_sampling_factor = cast(int)vertical_sampling
+					color_components[id].h_sampling_factor = cast(int)horizontal_sampling
+				}
+			case .SOF2: // Progressive DCT
+				unimplemented("SOF2")
+			case .SOF3: // Lossless (sequential)
+				fallthrough
+			case .SOF5: // Differential sequential DCT
+				fallthrough
+			case .SOF6: // Differential progressive DCT
+				fallthrough
+			case .SOF7: // Differential lossless (sequential)
+				fallthrough
+			case .SOF9: // Extended sequential DCT, Arithmetic coding
+				fallthrough
+			case .SOF10: // Progressive DCT, Arithmetic coding
+				fallthrough
+			case .SOF11: // Lossless (sequential), Arithmetic coding
+				fallthrough
+			case .SOF13: // Differential sequential DCT, Arithmetic coding
+				fallthrough
+			case .SOF14: // Differential progressive DCT, Arithmetic coding
+				fallthrough
+			case .SOF15: // Differential lossless (sequential), Arithmetic coding
+				fmt.println(marker)
+				return img, .Unsupported_Frame_Type
+			case .SOS:
+				if img.channels == 0 && img.depth == 0 && img.width == 0 && img.height == 0 {
+					return img, .Encountered_SOS_Before_SOF
+				}
+
+				if .do_not_decompress_image in options {
+					return img, nil
+				}
+
+				// Length
+				compress.read_data(ctx, u16be) or_return
+				num_components := compress.read_u8(ctx) or_return
+				if num_components != 1 && num_components != 3 {
+					return img, .Invalid_Number_Of_Channels
+				}
+
+				for _ in 0..<num_components {
+					component_id := cast(Component)compress.read_u8(ctx) or_return
+					if component_id < .Y || component_id > .Cr {
+						return img, .Image_Does_Not_Adhere_to_Spec
+					}
+
+					// high 4 is DC, low 4 is AC
+					coefficient_indices := compress.read_u8(ctx) or_return
+					dc_table_idx := coefficient_indices >> 4
+					ac_table_idx := coefficient_indices & 0xF
+
+					if (dc_table_idx < 0 || dc_table_idx > 3) || (ac_table_idx < 0 || ac_table_idx > 3) {
+						return img, .Invalid_Huffman_Table_Index
+					}
+
+					color_components[component_id].dc_table_idx = dc_table_idx
+					color_components[component_id].ac_table_idx = ac_table_idx
+				}
+				// TODO: These aren't used for sequential DCT, only progressive and lossless.
+				Ss := compress.read_u8(ctx) or_return
+				_ = Ss
+				Se := compress.read_u8(ctx) or_return
+				_ = Se
+				Ah_Al := compress.read_u8(ctx) or_return
+				_ = Ah_Al
+
+				blocks = make([]Block, block_height * block_width)
+
+				previous_dc: [Component]i16
+
+				restart_interval *= color_components[.Y].v_sampling_factor * color_components[.Y].h_sampling_factor
+				#no_bounds_check for y := 0; y < mcu_height; y += color_components[.Y].v_sampling_factor {
```

Comment:
Load the `color_components[.Y].v_sampling_factor` outside the for loop here for a small performance optimization. Odin cannot factor this out due to potential aliasing and modification of the `color_components` table above. In fact, I would hoist the load of all the `{v,h}_resampling_factor` for all components outside.